File: frontend/src/api/locations.ts
// src/api/locations.ts
import { api } from '@/lib/api';

export type Location = {
  id: string;
  source_type: string;
  source_id: string;
  raw_location_text: string;
  location_type?: string;
  latitude?: number;
  longitude?: number;
  geocode_confidence?: number;
  geocode_source?: string;
  street_name?: string;
  street_number?: string;
  city?: string;
  state?: string;
  postal_code?: string;
  country?: string;
  formatted_address?: string;
  playlist_uuid?: string;
  county_id?: number;
  geocoded_at?: string;
  created_at: string;
  updated_at?: string;
  // Context fields
  playlist_name?: string;
  county_name?: string;
  county_state?: string;
  transcript_text?: string;
  transcript_created_at?: string;
};

export type HeatmapPoint = {
  lat: number;
  lon: number;
  count: number;
  most_recent?: string;
};

export type LocationListResponse = {
  items: Location[];
  total: number;
  limit: number;
  offset: number;
};

export type HeatmapResponse = {
  points: HeatmapPoint[];
  total_locations: number;
  time_window_hours: number;
  center_lat?: number;
  center_lon?: number;
};

export type LocationStats = {
  total_locations: number;
  geocoded: number;
  unique_transcripts: number;
  unique_feeds: number;
  unique_cities: number;
  time_range: {
    oldest?: string;
    newest?: string;
  };
  time_window_hours: number;
};

export type LocationsQueryParams = {
  feed_id?: string;
  bbox?: string;
  since?: string;
  hours?: number;
  limit?: number;
  offset?: number;
};

export type HeatmapQueryParams = {
  feed_id?: string;
  hours?: number;
  grid_precision?: number;
};

const isMock = import.meta.env.VITE_MOCK === '1';

// Mock data for development
const mockLocations: Location[] = [
  {
    id: '1',
    source_type: 'transcript',
    source_id: 'call-1',
    raw_location_text: '123 Main Street',
    location_type: 'address',
    latitude: 33.2366,
    longitude: -96.8009,
    geocode_confidence: 0.9,
    city: 'Prosper',
    state: 'TX',
    formatted_address: '123 Main Street, Prosper, TX 75078',
    created_at: new Date().toISOString(),
    playlist_name: 'Prosper PD',
    transcript_text: 'Unit 12 responding to 123 Main Street.'
  },
  {
    id: '2',
    source_type: 'transcript',
    source_id: 'call-2',
    raw_location_text: '5th and Broadway',
    location_type: 'intersection',
    latitude: 33.2400,
    longitude: -96.7950,
    geocode_confidence: 0.85,
    city: 'Prosper',
    state: 'TX',
    formatted_address: '5th St & Broadway, Prosper, TX',
    created_at: new Date().toISOString(),
    playlist_name: 'Prosper PD',
    transcript_text: 'Accident reported at 5th and Broadway.'
  },
];

const mockHeatmapPoints: HeatmapPoint[] = [
  { lat: 33.2366, lon: -96.8009, count: 5, most_recent: new Date().toISOString() },
  { lat: 33.2400, lon: -96.7950, count: 3, most_recent: new Date().toISOString() },
  { lat: 33.2350, lon: -96.8050, count: 8, most_recent: new Date().toISOString() },
];

export async function getLocations(params?: LocationsQueryParams): Promise<LocationListResponse> {
  if (isMock) {
    return {
      items: mockLocations,
      total: mockLocations.length,
      limit: params?.limit ?? 500,
      offset: params?.offset ?? 0
    };
  }
  try {
    const response = await api.get<LocationListResponse>('/locations', { params });
    return response.data;
  } catch (error) {
    console.warn('Using mock locations due to API error', error);
    return {
      items: mockLocations,
      total: mockLocations.length,
      limit: params?.limit ?? 500,
      offset: params?.offset ?? 0
    };
  }
}

export async function getHeatmap(params?: HeatmapQueryParams): Promise<HeatmapResponse> {
  if (isMock) {
    return {
      points: mockHeatmapPoints,
      total_locations: mockHeatmapPoints.reduce((sum, p) => sum + p.count, 0),
      time_window_hours: params?.hours ?? 24,
      center_lat: 33.2370,
      center_lon: -96.8000
    };
  }
  try {
    const response = await api.get<HeatmapResponse>('/locations/heatmap', { params });
    return response.data;
  } catch (error) {
    console.warn('Using mock heatmap due to API error', error);
    return {
      points: mockHeatmapPoints,
      total_locations: mockHeatmapPoints.reduce((sum, p) => sum + p.count, 0),
      time_window_hours: params?.hours ?? 24,
      center_lat: 33.2370,
      center_lon: -96.8000
    };
  }
}

export async function getLocation(id: string): Promise<Location | undefined> {
  if (isMock) {
    return mockLocations.find(l => l.id === id);
  }
  try {
    const response = await api.get<Location>(`/locations/${id}`);
    return response.data;
  } catch (error) {
    console.warn('Using mock location due to API error', error);
    return mockLocations.find(l => l.id === id);
  }
}

export async function getLocationStats(params?: { feed_id?: string; hours?: number }): Promise<LocationStats> {
  if (isMock) {
    return {
      total_locations: 15,
      geocoded: 12,
      unique_transcripts: 10,
      unique_feeds: 2,
      unique_cities: 3,
      time_range: {
        oldest: new Date(Date.now() - 86400000).toISOString(),
        newest: new Date().toISOString()
      },
      time_window_hours: params?.hours ?? 24
    };
  }
  try {
    const response = await api.get<LocationStats>('/locations/stats/summary', { params });
    return response.data;
  } catch (error) {
    console.warn('Using mock stats due to API error', error);
    return {
      total_locations: 15,
      geocoded: 12,
      unique_transcripts: 10,
      unique_feeds: 2,
      unique_cities: 3,
      time_range: {},
      time_window_hours: params?.hours ?? 24
    };
  }
}

File: app_api/routers/locations.py
"""
Locations API Router

Endpoints for querying geocoded locations from police scanner transcripts.
Used for map visualization and heatmap features.
"""
from fastapi import APIRouter, Query, Depends, HTTPException
from typing import Optional, List
from datetime import datetime, timedelta, timezone
from uuid import UUID
import asyncpg

from database import get_pool
from models.locations import (
    Location, LocationWithContext, LocationListResponse,
    HeatmapResponse, HeatmapPoint
)

router = APIRouter()


def transform_location_row(row: dict) -> dict:
    """Transform database row to location dict."""
    result = dict(row)
    # Convert Decimal to float for JSON serialization
    for key in ['latitude', 'longitude', 'geocode_confidence']:
        if result.get(key) is not None:
            result[key] = float(result[key])
    return result


@router.get("", response_model=LocationListResponse)
async def list_locations(
    feed_id: Optional[UUID] = Query(None, description="Filter by playlist/feed UUID"),
    bbox: Optional[str] = Query(None, description="Bounding box: sw_lat,sw_lon,ne_lat,ne_lon"),
    since: Optional[datetime] = Query(None, description="Only locations since this timestamp"),
    hours: Optional[int] = Query(None, ge=1, le=720, description="Locations from last N hours"),
    limit: int = Query(500, ge=1, le=5000),
    offset: int = Query(0, ge=0),
    pool: asyncpg.Pool = Depends(get_pool)
):
    """
    List geocoded locations with optional filtering.

    Supports filtering by:
    - feed_id: Playlist UUID
    - bbox: Geographic bounding box (sw_lat,sw_lon,ne_lat,ne_lon)
    - since: Timestamp filter
    - hours: Time window in hours
    """
    query = """
        SELECT
            l.id, l.source_type, l.source_id, l.raw_location_text, l.location_type,
            l.latitude, l.longitude, l.geocode_confidence, l.geocode_source,
            l.street_name, l.street_number, l.city, l.state, l.postal_code,
            l.country, l.formatted_address, l.playlist_uuid, l.county_id,
            l.geocoded_at, l.created_at, l.updated_at,
            p.name as playlist_name,
            c.county_name, c.state_code as county_state,
            t.text as transcript_text, t.created_at as transcript_created_at
        FROM locations l
        LEFT JOIN bcfy_playlists p ON p.uuid = l.playlist_uuid
        LEFT JOIN bcfy_counties c ON c.cntid = l.county_id
        LEFT JOIN transcripts t ON t.call_uid = l.source_id AND l.source_type = 'transcript'
        WHERE l.latitude IS NOT NULL AND l.longitude IS NOT NULL
    """
    params = []
    param_count = 1

    # Filter by feed
    if feed_id:
        query += f" AND l.playlist_uuid = ${param_count}"
        params.append(feed_id)
        param_count += 1

    # Filter by bounding box
    if bbox:
        try:
            sw_lat, sw_lon, ne_lat, ne_lon = map(float, bbox.split(','))
            query += f" AND l.latitude BETWEEN ${param_count} AND ${param_count + 1}"
            query += f" AND l.longitude BETWEEN ${param_count + 2} AND ${param_count + 3}"
            params.extend([sw_lat, ne_lat, sw_lon, ne_lon])
            param_count += 4
        except (ValueError, AttributeError):
            raise HTTPException(status_code=400, detail="Invalid bbox format. Use: sw_lat,sw_lon,ne_lat,ne_lon")

    # Filter by time
    if hours:
        since_time = datetime.now(timezone.utc) - timedelta(hours=hours)
        query += f" AND l.created_at >= ${param_count}"
        params.append(since_time)
        param_count += 1
    elif since:
        query += f" AND l.created_at >= ${param_count}"
        params.append(since)
        param_count += 1

    # Get total count
    count_query = f"SELECT COUNT(*) FROM ({query}) subq"

    # Add ordering and pagination
    query += f" ORDER BY l.created_at DESC LIMIT ${param_count} OFFSET ${param_count + 1}"
    params.extend([limit, offset])

    async with pool.acquire() as conn:
        total = await conn.fetchval(count_query, *params[:-2])
        rows = await conn.fetch(query, *params)

    return LocationListResponse(
        items=[LocationWithContext(**transform_location_row(dict(row))) for row in rows],
        total=total or 0,
        limit=limit,
        offset=offset
    )


@router.get("/heatmap", response_model=HeatmapResponse)
async def get_heatmap(
    feed_id: Optional[UUID] = Query(None, description="Filter by playlist/feed UUID"),
    hours: int = Query(24, ge=1, le=720, description="Time window in hours"),
    grid_precision: int = Query(3, ge=2, le=5, description="Grid precision (decimals)"),
    pool: asyncpg.Pool = Depends(get_pool)
):
    """
    Get aggregated location data for heatmap visualization.

    Returns points aggregated by grid cells for efficient heatmap rendering.
    Grid precision controls cell size:
    - 2 = ~1km cells
    - 3 = ~100m cells (default)
    - 4 = ~10m cells
    - 5 = ~1m cells
    """
    since_time = datetime.now(timezone.utc) - timedelta(hours=hours)

    query = f"""
        SELECT
            ROUND(latitude::numeric, $1) as lat_grid,
            ROUND(longitude::numeric, $1) as lon_grid,
            AVG(latitude) as lat,
            AVG(longitude) as lon,
            COUNT(*) as count,
            MAX(created_at) as most_recent
        FROM locations
        WHERE latitude IS NOT NULL
          AND longitude IS NOT NULL
          AND created_at >= $2
    """
    params = [grid_precision, since_time]
    param_count = 3

    if feed_id:
        query += f" AND playlist_uuid = ${param_count}"
        params.append(feed_id)
        param_count += 1

    query += " GROUP BY ROUND(latitude::numeric, $1), ROUND(longitude::numeric, $1)"
    query += " HAVING COUNT(*) > 0"
    query += " ORDER BY count DESC"

    async with pool.acquire() as conn:
        rows = await conn.fetch(query, *params)

        # Get total count and center point
        stats = await conn.fetchrow("""
            SELECT
                COUNT(*) as total,
                AVG(latitude) as center_lat,
                AVG(longitude) as center_lon
            FROM locations
            WHERE latitude IS NOT NULL
              AND longitude IS NOT NULL
              AND created_at >= $1
        """ + (" AND playlist_uuid = $2" if feed_id else ""),
            *([since_time, feed_id] if feed_id else [since_time])
        )

    points = [
        HeatmapPoint(
            lat=float(row['lat']),
            lon=float(row['lon']),
            count=row['count'],
            most_recent=row['most_recent']
        )
        for row in rows
    ]

    return HeatmapResponse(
        points=points,
        total_locations=stats['total'] or 0,
        time_window_hours=hours,
        center_lat=float(stats['center_lat']) if stats['center_lat'] else None,
        center_lon=float(stats['center_lon']) if stats['center_lon'] else None
    )


@router.get("/{location_id}", response_model=LocationWithContext)
async def get_location(
    location_id: UUID,
    pool: asyncpg.Pool = Depends(get_pool)
):
    """Get a single location with full context."""
    query = """
        SELECT
            l.id, l.source_type, l.source_id, l.raw_location_text, l.location_type,
            l.latitude, l.longitude, l.geocode_confidence, l.geocode_source,
            l.street_name, l.street_number, l.city, l.state, l.postal_code,
            l.country, l.formatted_address, l.playlist_uuid, l.county_id,
            l.geocoded_at, l.created_at, l.updated_at,
            p.name as playlist_name,
            c.county_name, c.state_code as county_state,
            t.text as transcript_text, t.created_at as transcript_created_at
        FROM locations l
        LEFT JOIN bcfy_playlists p ON p.uuid = l.playlist_uuid
        LEFT JOIN bcfy_counties c ON c.cntid = l.county_id
        LEFT JOIN transcripts t ON t.call_uid = l.source_id AND l.source_type = 'transcript'
        WHERE l.id = $1
    """

    async with pool.acquire() as conn:
        row = await conn.fetchrow(query, location_id)

    if not row:
        raise HTTPException(status_code=404, detail="Location not found")

    return LocationWithContext(**transform_location_row(dict(row)))


@router.get("/stats/summary")
async def get_location_stats(
    feed_id: Optional[UUID] = Query(None, description="Filter by playlist/feed UUID"),
    hours: int = Query(24, ge=1, le=720, description="Time window in hours"),
    pool: asyncpg.Pool = Depends(get_pool)
):
    """Get summary statistics for locations."""
    since_time = datetime.now(timezone.utc) - timedelta(hours=hours)

    query = """
        SELECT
            COUNT(*) as total_locations,
            COUNT(*) FILTER (WHERE latitude IS NOT NULL) as geocoded,
            COUNT(DISTINCT source_id) as unique_transcripts,
            COUNT(DISTINCT playlist_uuid) as unique_feeds,
            COUNT(DISTINCT city) FILTER (WHERE city IS NOT NULL) as unique_cities,
            MIN(created_at) as oldest,
            MAX(created_at) as newest
        FROM locations
        WHERE created_at >= $1
    """
    params = [since_time]

    if feed_id:
        query += " AND playlist_uuid = $2"
        params.append(feed_id)

    async with pool.acquire() as conn:
        row = await conn.fetchrow(query, *params)

    return {
        "total_locations": row['total_locations'],
        "geocoded": row['geocoded'],
        "unique_transcripts": row['unique_transcripts'],
        "unique_feeds": row['unique_feeds'],
        "unique_cities": row['unique_cities'],
        "time_range": {
            "oldest": row['oldest'].isoformat() if row['oldest'] else None,
            "newest": row['newest'].isoformat() if row['newest'] else None
        },
        "time_window_hours": hours
    }
